File: Lesson 7, Stacks and Queues/Lesson 7, Stacks and Queues, Fish.py
def solution(A, B):
    i, alive, N, stack = 0, len(A), len(A), []
    while i < N:
        if B[i]:
            stack.append(i)
        elif stack:
            alive -= 1
            if A[stack[-1]] < A[i]:
                stack.pop(); i -= 1
        i += 1
    return alive

File: Lesson 7, Stacks and Queues/Lesson 7, Stacks and Queues, Fish.js
function solution(A, B) {
    let i = 0, alive = A.length, N = A.length, stack = [] 
    for (; i < N; i++) {
        if (parseInt(B[i]) === 1)
            stack.push(i)
        else if (stack.length > 0) {
            alive--
            if (parseInt(A[stack.length - 1]) < parseInt(A[i])) {
                stack.pop()
                i--
            }
        }
    }
    return alive
}